### PR TITLE
Support OBS video playback

### LINKED ISSUE #655: Solve the OBS video scheduling issues
From https://github.com/srobo/competition-team-minutes/blob/master/SR2021/2021-02-06-league-2-retro.md#specific

Relevant slack thread: https://studentrobotics.slack.com/archives/C3190DWP3/p1612024768005400

Currently we use https://github.com/RealOrangeOne/livestream-video-scheduler to play the match videos, however this doesn't work in the OBS internal browser on Linux due to it not having the necessary codecs.

There are various alternatives for playing the videos at the proper times, the most promising seem:
- OBS supports scripting
- the OBS websocket extension

### PR DESCRIPTION
**Pull request by PeterJCLaw**
This does some further refactoring to simplify things and then introduces support for telling OBS to play videos via [obs-websocket-py](https://pypi.org/project/obs-websocket-py/).

This itself isn't tested, though I have previously tested OBS websocket usage like this in https://github.com/PeterJCLaw/obs-match-scheduler/blob/main/via-obs-websocket-py.

Fixes https://github.com/srobo/tasks/issues/655.

### CODE REVIEW

File: sr/comp/mixtape/mixtape.py [hunk 1]
```diff
@@ -23,12 +24,33 @@ def __init__(
...
         self.magicq_controller = magicq_controller
+        self.obs_studio_controller = obs_studio_controller
+
+    def get_play_video_action(
+        self,
+        track: Any,
+        current_offset: Callable[[], float],
+    ) -> Tuple[Action, str]:
+        path = os.path.join(self.root, track['obs_video'])
+        preload(path)
```
Comment (WillB97) [inline sr/comp/mixtape/mixtape.py:42]: I don't think we should be attempting to game the cache. Being a production software OBS should do sufficient caching on media sources so we should just let it handle it's own media. Also being a recording, the video doesn't actually have to sync with anything so the latency is not super critical.

File: sr/comp/mixtape/mixtape.py [hunk 2]
```diff
@@ -23,12 +24,33 @@ def __init__(
...
         self.exclusivity_groups: Dict[object, subprocess.Popen[bytes]] = {}
         self.magicq_controller = magicq_controller
+        self.obs_studio_controller = obs_studio_controller
+
+    def get_play_video_action(
+        self,
+        track: Any,
+        current_offset: Callable[[], float],
+    ) -> Tuple[Action, str]:
+        path = os.path.join(self.root, track['obs_video'])
```
Comment (WillB97) [inline sr/comp/mixtape/mixtape.py:41]: By converting this into an absolute path we are forcing mixtape to run on the same computer as OBS. Ultimately leaving this as an OBS-relative path gives us the option to separate the services.

File: sr/comp/mixtape/obs_studio.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+            websocket.call(requests.SetSceneItemProperties(
+                {'name': self.source_name},
+                # Setting the bounds here means that the media will fill the canvas,
+                # preserving its aspect ratio.
+                bounds={
+                    'type': 'OBS_BOUNDS_SCALE_INNER',
+                    'x': self.video_info.getBaseWidth(),
+                    'y': self.video_info.getBaseHeight(),
+                },
+            ))
```
Comment (WillB97) [inline sr/comp/mixtape/obs_studio.py:83]: Do we actually need to update the size each time since it hasn't changed?

**Review by WillB97 — COMMENTED** [review]
Great work on this. No problems with the implementation however I'm concerned that the current setup overly restricts us to using a singular machine running both mixtape and OBS while the use of websockets enables the option of these being separate.

**Comment by PeterJCLaw** [comment]
> I'm concerned that the current setup overly restricts us to using a singular machine running both mixtape and OBS while the use of websockets enables the option of these being separate.

I'm not sure why this would be a problem, nor why you'd want to run mixtape on another machine. Perhaps I should remove the hostname configuration option to clarify that that's not an expected use-case? If we were to run mixtape on another machine that would likely introduce latency and security issues, both of which are better off avoided.

I really don't view the use of websockets here as a way to decouple OBS from mixtape, but rather only as a convenient interface.

File: sr/comp/mixtape/obs_studio.py (hunk 3 above)
Comment (PeterJCLaw) [reply sr/comp/mixtape/obs_studio.py:83]: Yes. OBS seems to play videos at their native resolution by default, which we're overriding here. I'll add a comment.

File: sr/comp/mixtape/mixtape.py (hunk 1 above)
Comment (PeterJCLaw) [reply sr/comp/mixtape/mixtape.py:42]: The video needs to sync with the overlay, at least to some extent.

With the current way that this works, OBS cannot do any preloading itself as it doesn't know what file to play until it's told to actually play it. I did previously explore various things to tell OBS about a video file before playing it, however there didn't seem to be a good mechanism for doing so -- everything I tried resulted in OBS playing the file immediately.

File: sr/comp/mixtape/obs_studio.py [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+        self.source_name = source
+        self.video_info = websocket.call(requests.GetVideoInfo())
+
+        # Our play_video method is going to be called from one of the (possibly
+        # many) worker threads which the scheduler will use. Guard it against
+        # concurrent access.
+        self.websocket = Guarded(websocket)
+
+    def play_video(self, filename: str) -> None:
+        with self.websocket as websocket:
```
Comment (PeterJCLaw) [inline sr/comp/mixtape/obs_studio.py:63]: Hrm, just realised that I'm not actually doing the transition to the scene here. We should probably do that too.

**Comment by WillB97** [comment]
__Setup notes:__
For the source being controlled the option `Close file when inactive` needs to be set to allow the source to be changed when not active.
The scene being transitioned to needs `Transition Override > Fade` selected so there is a fade

**Review by sedders123 — APPROVED** [review]

File: sr/comp/mixtape/mixtape.py (hunk 2 above)
Comment (PeterJCLaw) [reply sr/comp/mixtape/mixtape.py:41]: Have clarified this by removing the host option.

File: sr/comp/mixtape/obs_studio.py (hunk 4 above)
Comment (PeterJCLaw) [reply sr/comp/mixtape/obs_studio.py:63]: https://github.com/srobo/srcomp-mixtape/pull/8/commits/a821901efd9aef28abe6b08a65cbd58680951e90 and https://github.com/srobo/srcomp-mixtape/pull/8/commits/b83dc66b2588d35a0aefa992a0bc901e597caa4c do now do this.

**Pull request merged**